kuma/wiki/tests/test_jobs.py

```diff
@@ -1,141 +1,86 @@
-                         created=rev_2.created + timedelta(seconds=1),
-                         save=True)
-        ok_(rev_2.created < rev_3.created)
-        job_user_pks = [contributor['id']
-                        for contributor in job.fetch(rev_1.document.pk)]
-        # The new revision shows up
-        eq_(job_user_pks, [contrib_3.pk] + recent_contributors_pks)
-
-    def test_contributors_inactive_or_banned(self):
-        contrib_1 = user(save=True)
-        contrib_2 = user(is_active=False, save=True)
-        contrib_3 = user(save=True)
-        contrib_3_ban = contrib_3.bans.create(by=contrib_1, reason='because reasons')
-        revision_2 = revision(creator=contrib_1, save=True)
-
-        revision(creator=contrib_2, document=revision_2.document, save=True)
-        revision(creator=contrib_3, document=revision_2.document, save=True)
-
-        job = DocumentContributorsJob()
-        # setting this to true to be able to test this
-        job.fetch_on_miss = True
-
-        contributors = job.get(revision_2.document.pk)
-        contrib_ids = [contrib['id'] for contrib in contributors]
-        self.assertIn(contrib_1.id, contrib_ids)
-        self.assertNotIn(contrib_2.id, contrib_ids)
-        self.assertNotIn(contrib_3.id, contrib_ids)
-
-        # delete the ban again
-        contrib_3_ban.delete()
-        # reloading the document from db to prevent cache
-        doc = Document.objects.get(pk=revision_2.document.pk)
-        # user not in contributors because job invalidation hasn't happened
-        contrib_ids = [contrib['id']
-                       for contrib in job.get(revision_2.document.pk)]
-        self.assertNotIn(contrib_3.id, contrib_ids)
-
-        # trigger the invalidation manually by saving the document
-        doc.save()
-        doc = Document.objects.get(pk=revision_2.document.pk)
-        contrib_ids = [contrib['id']
-                       for contrib in job.get(revision_2.document.pk)]
-        self.assertIn(contrib_3.id, contrib_ids)
+def test_find_roots(db_and_empty_caches, multi_generational_docs, root_doc):
+    """
+    Ensure sub pages can find the content zone root.
+    """
+    def get_zone_stack(doc):
+        return DocumentZoneStackJob().get(doc.pk)
+
+    top_doc = multi_generational_docs.great_grandparent
+    middle_doc = multi_generational_docs.grandparent
+    below_middle_doc = multi_generational_docs.parent
+    bottom_doc = multi_generational_docs.child
+
+    top_zone = DocumentZone(document=top_doc)
+    top_zone.save()
+
+    middle_zone = DocumentZone(document=middle_doc)
+    middle_zone.save()
+
+    assert unicode(top_zone) == u'DocumentZone {} ({})'.format(
+        top_doc.get_absolute_url(), top_doc.title)
+
+    assert get_zone_stack(top_doc) == [top_zone]
+    assert get_zone_stack(middle_doc) == [middle_zone, top_zone]
+    assert get_zone_stack(below_middle_doc) == [middle_zone, top_zone]
+    assert get_zone_stack(bottom_doc) == [middle_zone, top_zone]
```

Comment:
Nice rewrite, and I prefer this to how ``DocumentZoneTests.test_find_roots`` just tested the first element.